### PR TITLE
DynamoDB CDK for Archive Parts Table

### PR DESCRIPTION
**Pull request by dbb-aws-only**
https://app.asana.com/0/1202938635037719/1203922468885149/f

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/refreezer/infrastructure/stack.py [hunk 1]
```diff
@@ -73,6 +74,25 @@ def __init__(
             value=table.table_name,
         )
 
+        archive_parts_table = dynamodb.Table(
+            self,
+            "ArchivePartsTable",
```
Comment (bassemwanis) [inline source/refreezer/infrastructure/stack.py:79]: If we will use this table for Inventory tracking as well, should we have a more generic name?

File: source/refreezer/infrastructure/stack.py [hunk 2]
```diff
@@ -73,6 +74,25 @@ def __init__(
...
+            partition_key=dynamodb.Attribute(
+                name="archive_id", type=dynamodb.AttributeType.STRING
+            ),
+            sort_key=dynamodb.Attribute(
+                name="part_number", type=dynamodb.AttributeType.NUMBER
+            ),
+            point_in_time_recovery=True,
+            stream=dynamodb.StreamViewType.NEW_AND_OLD_IMAGES,
```
Comment (bassemwanis) [inline source/refreezer/infrastructure/stack.py:87]: Do we need stream for this table?

**Review by bassemwanis — COMMENTED** [review]
Once confirming the final schema, we will need unit and integration tests.

File: source/refreezer/infrastructure/stack.py (hunk 1 above)
Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py:79]: Agreed - let's change it once we understand what the final schema for archive will be and what we need for Inventory.

File: source/refreezer/infrastructure/stack.py (hunk 2 above)
Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py]: Nope - good call.  We'll let the Lambda pull once the job is complete.   It's been removed.

**Comment by kamyarz-aws** [comment]
I think you need to add a test to test_stack.py to check if table is created in cdk template

**Comment by dbb-aws-only** [comment]
> I think you need to add a test to test_stack.py to check if table is created in cdk template

The table only shows up when you attach it to a Lambda.  I'll update the tests once the table is in use.

File: source/refreezer/infrastructure/stack.py [hunk 3]
```diff
@@ -76,6 +77,22 @@ def __init__(
...
         )
 
+        archive_retrieval_table = dynamodb.Table(
+            self,
+            "GlacierArchiveRetrieval",
+            partition_key=dynamodb.Attribute(
+                name="pk", type=dynamodb.AttributeType.STRING
+            ),
+            sort_key=dynamodb.Attribute(name="sk", type=dynamodb.AttributeType.STRING),
+            point_in_time_recovery=True,
```
Comment (schana) [inline source/refreezer/infrastructure/stack.py:87]: Why is this enabled?

File: source/refreezer/infrastructure/stack.py [hunk 4]
```diff
@@ -76,6 +77,22 @@ def __init__(
             value=table.table_name,
         )
 
+        archive_retrieval_table = dynamodb.Table(
```
Comment (schana) [inline source/refreezer/infrastructure/stack.py:80]: This needs tested.

**Review by schana — CHANGES_REQUESTED** [review]
The commits shouldn't include the merge commit. Please clean this up as well.

**Comment by dbb-aws-only** [comment]
Closed in error

File: source/refreezer/infrastructure/stack.py [hunk 5]
```diff
@@ -29,6 +29,7 @@
 
 
 class OutputKeys:
+    ARCHIVE_RETRIEVAL_TABLE_NAME = "ARTN"
```
[the remark this answers is not in the document]
Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py]: Done.  I also renamed the variable to reflect the fact that we're also storing Inventory in addition to Archive data.

File: source/refreezer/infrastructure/stack.py [hunk 6]
```diff
@@ -76,6 +77,22 @@ def __init__(
             value=table.table_name,
         )
 
+        archive_retrieval_table = dynamodb.Table(
+            self,
+            "GlacierArchiveRetrieval",
```
[the remark this answers is not in the document]
Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py]: Renamed.

File: source/refreezer/infrastructure/stack.py (hunk 3 above)
Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py]: Removed.

Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py:87]: Actually - there's a CDK nag saying that the Point In Time is more insurance...

`The DynamoDB table does not have Point-in-time Recovery enabled. DynamoDB continuous backups represent an additional layer of insurance against accidental loss of data on top of on-demand backups. The DynamoDB service can back up the data with per-second granularity and restore it to any single second from the time PITR was enabled up to the prior 35 days.\n`

**Comment by dbb-aws-only** [comment]
> The commits shouldn't include the merge commit. Please clean this up as well.

Done

File: source/refreezer/infrastructure/stack.py (hunk 3 above)
Comment (schana) [reply source/refreezer/infrastructure/stack.py:88]: This is a cost consideration. We should leave it disabled and suppress the cdk nag, and add a task to the backlog to address later.

Comment (dbb-aws-only) [reply source/refreezer/infrastructure/stack.py:88]: Ok - I'll add the Nag Suppression code.  I'll add an Asana Task, too.

**Review by kamyarz-aws — APPROVED** [review]

File: source/refreezer/pipeline/stack.py [hunk 7]
```diff
@@ -124,6 +124,20 @@ def get_integration_test_step(self) -> pipelines.CodeBuildStep:
...
+                iam.PolicyStatement(
+                    effect=iam.Effect.ALLOW,
+                    actions=[
+                        "dynamodb:GetItem",
+                        "dynamodb:PutItem",
+                        "dynamodb:DeleteItem",
+                    ],
+                    resources=[
+                        (
+                            f"arn:aws:dynamodb:{Aws.REGION}:{Aws.ACCOUNT_ID}:table/"
```
Comment (bassemwanis) [inline source/refreezer/pipeline/stack.py:136]: Can you re-use same PolicyStatement of `AsyncFacilitatorTable` and add `GlacierObjectRetrieval` table to the resources array?

Comment (dbb-aws-only) [reply source/refreezer/pipeline/stack.py]: Done.

**Comment by dbb-aws-only** [comment]
> Once confirming the final schema, we will need unit and integration tests.

They are both added now - Simon and I are making very minor changes to the schema as we go, but the bulk of the I/O is well understood now.  There are both Unit and Integration tests in this PR now.

**Review by bassemwanis — APPROVED** [review]

**Pull request merged**